modules/firebase/firebase.go

```diff
@@ -0,0 +1,165 @@
+		m := testcontainers.ContainerMount{
+			Source: testcontainers.DockerVolumeMountSource{
+				Name:          volumeName,
+				VolumeOptions: volumeOptions,
+			},
+			Target: cacheFilePath,
+		}
+		req.Mounts = append(req.Mounts, m)
+		return nil
+	}
+}
+
+func gatherPorts(config FirebaseConfig) ([]string, error) {
+	var ports []string
+
+	v := reflect.ValueOf(config.Emulators)
+	for i := 0; i < v.NumField(); i++ {
+		emulator := v.Field(i)
+		if emulator.Kind() != reflect.Struct {
+			continue
+		}
+		name := v.Type().Field(i).Name
+
+		enabledF := emulator.FieldByName("Enabled")
+		if enabledF != (reflect.Value{}) && !enabledF.Bool() {
+			continue
+		}
+
+		hostF := emulator.FieldByName("Host")
+		portF := emulator.FieldByName("Port")
+		websocketPortF := emulator.FieldByName("WebsocketPort")
+
+		if hostF != (reflect.Value{}) && !hostF.IsZero() {
+			host := hostF.String()
+			if host != "0.0.0.0" {
+				return nil, fmt.Errorf("config specified %s emulator host on non public ip: %s", name, host)
+			}
+		}
+		if portF != (reflect.Value{}) && !portF.IsZero() {
+			port := fmt.Sprintf("%d/tcp", portF.Int())
+			ports = append(ports, port)
+		}
+		if websocketPortF != (reflect.Value{}) && !websocketPortF.IsZero() {
+			port := fmt.Sprintf("%d/tcp", websocketPortF.Int())
+			ports = append(ports, port)
+		}
+	}
+
+	return ports, nil
+}
+
+// Run creates an instance of the Firebase container type
+func Run(ctx context.Context, img string, opts ...testcontainers.ContainerCustomizer) (*Container, error) {
+	req := testcontainers.GenericContainerRequest{
+		ContainerRequest: testcontainers.ContainerRequest{
+			Image:      img,
+			Env:        map[string]string{},
+			WaitingFor: wait.ForLog("All emulators ready! It is now safe to connect your app."),
+		},
+		Started: true,
+	}
+
+	for _, opt := range opts {
+		if err := opt.Customize(&req); err != nil {
+			return nil, fmt.Errorf("customize: %w", err)
+		}
+	}
+
+	// Check if user supplied root:
+	rootPathIdx := slices.IndexFunc(req.ContainerRequest.Files, func(file testcontainers.ContainerFile) bool {
+		return file.ContainerFilePath == rootFilePath
+	})
+	if rootPathIdx == -1 {
+		return nil, fmt.Errorf("firebase emulator is unable to boot without configuration root (obtained via 'firebase init')")
+	}
+	// Parse expected emulators from the root:
+	userRoot := req.ContainerRequest.Files[rootPathIdx].HostFilePath
+	cfg, err := os.Open(path.Join(userRoot, "firebase.json"))
+	if err != nil {
+		return nil, fmt.Errorf("open firebase.json: %w", err)
+	}
+	defer cfg.Close()
+	bytes, err := io.ReadAll(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("read firebase.json: %w", err)
+	}
+	var parsed FirebaseConfig
+	if err := json.Unmarshal(bytes, &parsed); err != nil {
+		return nil, fmt.Errorf("parse firebase.json: %w", err)
+	}
+	expectedExposedPorts, err := gatherPorts(parsed)
+	if err != nil {
+		return nil, fmt.Errorf("issues with emulator config block: %w", err)
```

Comment:
suggestion: seems like this is just an invalid config
```suggestion
		return nil, fmt.Errorf("invalid config %q: %w", configPath, err)
```